Use 127.0.0.1 in dev and test

chulkilee opened the pull request:
When `:ip` is not given, phoenix is listening on all interfaces. This is not desired for dev/test mode, where it's unlikely to expose servers to public interfaces.

Add `:ip` option with ipv4 loopback (`127.0.0.1`) to all generated dev/test configuration.

(kind of related: #3400)

nathanl commented:
I think this is a good idea - I accidentally created a dup in https://github.com/phoenixframework/phoenix/pull/4010 but closed it after I saw this PR.

I think some explanatory comments would be helpful - see my PR and feel free to use what I wrote, or I can PR some comments afterward if this PR is merged.

chulkilee commented:
@nathanl thanks a lot! I copied some from your PR #4010 . Please feel free to make any suggestion on the help text.

chrismccord commented:
Looks like #3400 covers this. We like the idea, but need to sanity check a few things around windows iirc. If we go this direction, it will land along with 1.5. Thanks!!

chulkilee commented:
@chrismccord #3400 *does not* cover this, since it only updates `installer/templates/phx_single/config/prod.secret.exs`  while this PR is only to update  *dev* and *test*.

Can we reopen this ticket for dev/test change?

chulkilee commented:
I also confirmed it works on Windows with the follow step:

- Downloaded virtualbox vm "MSEdge on Win10 (x64) Stable 1809" from https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/
- Installed Elixir and Erlang with the installer from https://elixir-lang.org/install.html#windows
  - Elixir 1.10.4, Erlang 21.3 (default at this point)
- Added `http: [ip: { 127, 0, 0, 1 }, port: 4000]` to `dev.exs`
- Started the server with `mix phx.server` => it showed `localhost:4000`  instead of `0.0.0.0:4000`
  - Also it didn't ask for the firewall (it happened when I dropped the config to use `0.0.0.0:4000`)
- Opened http://127.0.0.1:4000 from IE => it worked!

nathanl commented on installer/templates/phx_single/config/dev.exs:
```diff
@@ -7,7 +7,9 @@ import Config
 config :<%= app_name %>, <%= endpoint_module %>,
-  http: [port: 4000],
+  # Binding to loopback ipv4 address prevents access from other machines.
+  # Change to `ip: {0, 0, 0, 0}` to allow access from other machines.
```
After "to allow access from other machines", I would add "or from the host machine if Phoenix is running in a VM or container""

chulkilee replied:
Running in vm / container does not necessarily requires binding on public interface - e.g. users may run proxy on the machine - in that case it runs in container/vm but does require that... but I'd like to add that as an "example" (not another case with "or")  :heart:.

Could you help improve the wording?

> Change to `ip: {0, 0, 0, 0}` to allow access from other machines,
> such as accessing the app runnining in a virtual machine or container
> from the host machine if Phoenix is running in a VM or container

Also.. I'm not sure "Phoenix is running" is right statement. Do we have similar comments somewhere? We're running this app, not "Phoenix".

nathanl commented:
@chrismccord Just seconding that this is not the same as #3400 - that's about prod, this is about dev and test.

chulkilee commented:
@chrismccord @josevalim could you reopen and review this PR? As @nathanl said, this PR is different from #3400 - and actually #3993 (with IP address tuple) was merged by @josevalim .

I rebased the branch to fix conflict. If needed I'll create new PR - let me know!
